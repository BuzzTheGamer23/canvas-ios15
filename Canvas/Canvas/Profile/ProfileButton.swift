//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
    
    

import UIKit
import Peeps
import ReactiveSwift
import TooLegit
import TechDebt
import Kingfisher

func addProfileButton(_ session: Session, viewController: UIViewController) {
    let profileButton = ProfileBarButtonItem(avatarURL: session.user.avatarURL)
    let enrollments = viewController
    
    profileButton.rac_command = RACCommand() { [weak profileButton, weak enrollments] _ in
        guard let profileButton = profileButton, let enrollments = enrollments else { return .empty() }
        let profile = ProfileViewController()
        profile.settingsViewControllerFactory = {
            return SettingsViewController.controller(CKCanvasAPI.current())
        }
        profile.canvasAPI = CKCanvasAPI.current()
        profile.user = profile.canvasAPI.user
        profile.profileImageSelected = { newProfileImage in
            if let key = session.user.avatarURL?.absoluteString {
                if let image = newProfileImage {
                    KingfisherManager.shared.cache.store(image, forKey: key)
                } else {
                    KingfisherManager.shared.cache.removeImage(forKey: key)
                }
            }
            
            profileButton.setProfileImage(newProfileImage)
        }
        let doneButton = UIBarButtonItem(barButtonSystemItem: .done, target: nil, action: nil)
        doneButton.rac_command = RACCommand() { [weak profile] _ in
            profile?.dismiss(animated: true, completion: nil)
            return .empty()
        }
        profile.navigationItem.leftBarButtonItem = doneButton
        let nav = UINavigationController(rootViewController: profile)
        nav.modalPresentationStyle = .formSheet
        enrollments.present(nav, animated: true, completion: nil)
        return .empty()
    }
    enrollments.navigationItem.leftBarButtonItem = profileButton
}
