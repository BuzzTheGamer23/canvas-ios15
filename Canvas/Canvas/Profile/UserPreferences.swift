//
//  UserPreferences.swift
//  iCanvas
//
//  Created by Derrick Hathaway on 3/22/16.
//  Copyright Â© 2016 Instructure. All rights reserved.
//

import Foundation
import UIKit


enum UserPreferences {
    
    enum LandingPage: String {
        case Courses        = "Courses"
        case Calendar       = "Calendar"
        case ToDo           = "To-Do List"
        case Notifications  = "Notifications"
        case Messages       = "Messages"

        /// Note: this must be synchronized with the list in RootViewController()
        var tabIndex: Int {
            switch self {
            case .Courses:          return 0
            case .Calendar:         return 1
            case .ToDo:             return 2
            case .Notifications:    return 3
            case .Messages:         return 4
            }
        }
    }
    
    private static let LandingPageKey = "landingPageSettings"
    
    static func landingPage(userID: String) -> UserPreferences.LandingPage {
        guard let landingPagePreferencesByUser: [String: String] = NSUserDefaults.standardUserDefaults()
            .objectForKey(UserPreferences.LandingPageKey) as? [String: String] else {
                return .Courses
        }
        
        return landingPagePreferencesByUser[userID]
            .flatMap(LandingPage.init)
            ?? .Courses
    }
}


