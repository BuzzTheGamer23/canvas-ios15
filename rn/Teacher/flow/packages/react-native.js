// @flow
export type PanResponderGestureState = {
  stateID: number,
  moveX: number,
  moveY: number,
  x0: number,
  y0: number,
  dx: number,
  dy: number,
  vx: number,
  vy: number,
}
