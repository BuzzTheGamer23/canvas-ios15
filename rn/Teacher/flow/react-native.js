//
// Copyright (C) 2017-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// @flow
/* eslint-disable */

// define basic style prop types
// the internal ones are the best we have
import {
  type ____ViewStyleProp_Internal,
  type ____TextStyleProp_Internal,
  type ____ImageStyleProp_Internal,
} from 'react-native/Libraries/StyleSheet/StyleSheetTypes'

declare type ViewStyleProp = ____ViewStyleProp_Internal
declare type TextStyleProp = ____TextStyleProp_Internal
declare type ImageStyleProp = ____ImageStyleProp_Internal
