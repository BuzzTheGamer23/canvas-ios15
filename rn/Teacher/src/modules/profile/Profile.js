//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

/* @flow */

import React, { Component } from 'react'
import {
  View,
  NativeModules,
  StyleSheet,
  Image,
  Dimensions,
  ActionSheetIOS,
  Linking,
  AlertIOS,
  TouchableWithoutFeedback,
} from 'react-native'
import i18n from 'format-message'
import { Text, Heading1, Paragraph } from '../../common/text'
import Avatar from '../../common/components/Avatar'
import Screen from '../../routing/Screen'
import Images from '../../images'
import branding from '../../common/branding'
import color from '../../common/colors'
import device from 'react-native-device-info'

const { width: deviceWidth } = Dimensions.get('window')

import { getSession } from '../../canvas-api'

export default class Profile extends Component {

  constructor (props: any) {
    super(props)
    this.secretTapCount = 0

    const settingsActions = [
      { title: i18n('Visit the Canvas Guides'), id: 'canvas-guides' },
    ]
    if (NativeModules.RNMail.canSendMail) {
      settingsActions.push({ title: i18n('Report a Problem'), id: 'mail' })
    }

    settingsActions.push({ title: i18n('Terms of Use'), id: 'terms' })
    settingsActions.push({ title: i18n('Cancel'), id: 'cancel' })
    this.settingsActions = settingsActions
  }

  logout = () => {
    NativeModules.NativeLogin.logout()
  }

  secretTap = () => {
    this.secretTapCount++
    if (this.secretTapCount > 10) {
      this.secretTapCount = 0
      this.props.navigator.show('/staging', { modal: true })
    }
  }

  settings = () => {
    ActionSheetIOS.showActionSheetWithOptions({
      options: this.settingsActions.map(a => a.title),
      cancelButtonIndex: this.settingsActions.length - 1,
    }, this.handleActions)
  }

  handleActions = (index: number) => {
    const action = this.settingsActions[index]
    switch (action.id) {
      case 'canvas-guides':
        Linking.openURL('https://community.canvaslms.com/community/answers/guides/mobile-guide/content?filterID=contentstatus%5Bpublished%5D~category%5Btable-of-contents%5D')
        break
      case 'mail':
        NativeModules.RNMail.mail({
          subject: 'Canvas Teacher - Report a problem',
          recipients: ['support@instructure.com'],
        }, (error, event) => {
          if (error) {
            AlertIOS.alert(i18n('Unavailable'), i18n('Mail is unavailable on your device.'))
          }
        })
        break
      case 'terms':
        Linking.openURL('https://www.canvaslms.com/policies/terms-of-use')
        break
      default:
        break
    }
  }

  render () {
    const session = getSession()

    if (!session) return null
    const user = session.user
    return (
      <Screen
        navBarHidden={false}
        navBarImage={branding.headerImage}
        navBarButtonColor={color.navBarTextColor}
        navBarColor={color.navBarColor}
        navBarStyle='dark'
        leftBarButtons={[
          {
            title: i18n('Done'),
            testID: 'profile.navigation-dismiss-btn',
            action: this.props.navigator.dismiss,
            accessibilityLabel: i18n('Done'),
          },
        ]}
        rightBarButtons={[
          {
            title: i18n('Log Out'),
            testID: 'profile.navigation-logout-btn',
            action: this.logout,
          },
          {
            image: Images.course.settings,
            testID: 'profile.navigation-settings-btn',
            action: this.settings,
            accessibilityLabel: i18n('Profile actions'),
          },
        ]}
      >
        <View testID="module.profile" accessible={true} style={styles.container}>
          <View style={styles.headerImageContainer} >
            <Image source={Images.pencilBG} style={styles.headerImage} />
          </View>
          <View style={styles.profileImageContainer}>
            <Avatar
              avatarURL={user.avatar_url}
              userName={user.name}
              border={true}
              height={160} />
          </View>
          <View style={styles.profileInfoContainer}>
            <Heading1>{user.name}</Heading1>
            <Paragraph>{user.primary_email}</Paragraph>
          </View>
          <View style={styles.versionContainer}>
            <TouchableWithoutFeedback onPress={this.secretTap} testID='profile-btn-secret-tap'>
              <Text style={styles.versionText}>{i18n('Version {version}', { version: device.getReadableVersion() })}</Text>
            </TouchableWithoutFeedback>
          </View>
        </View>
      </Screen>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  topContainer: {
    backgroundColor: '#394B58',
    height: 64,
  },
  profileInfoContainer: {
    top: -80,
    alignItems: 'center',
  },
  headerImageContainer: {
    backgroundColor: 'lightgrey',
  },
  headerImage: {
    height: 170,
    width: deviceWidth,
  },
  profileImageContainer: {
    top: -80,
    marginBottom: 20,
    flex: 0,
    flexDirection: 'row',
    justifyContent: 'center',
  },
  versionContainer: {
    bottom: -125,
    alignItems: 'center',
  },
  versionText: {
    color: '#73818C',
    fontSize: 12,
  },
})
