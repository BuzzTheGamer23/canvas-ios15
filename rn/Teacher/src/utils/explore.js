/* @flow */

function hasProp (object: any, key: string, value: any): boolean {
  if (object instanceof Object) {
    const props = object.props
    if (props && props instanceof Object) {
      for (var prop in props) {
        if (prop === key && props[prop] === value) {
          return true
        }
      }
    }
  }

  return false
}

function selectByProp (object: any, propKey: string, value: any): ?any {
  if (!(typeof object === 'object')) {
    return null
  }

  if (hasProp(object, propKey, value)) {
    return object
  }

  let result = null

  for (var key in object) {
    if (object[key] instanceof Array) {
      object[key].forEach((e) => {
        result = selectByProp(e, propKey, value)
        if (result) {
          return result
        }
      })
    }

    result = selectByProp(object[key], propKey, value)
    if (result) {
      return result
    }
  }

  return result
}

class ComponentExplorer {
  component: any

  constructor (component: any) {
    this.component = component
  }

  selectByProp (propKey: string, value: any): ?any {
    return selectByProp(this.component, propKey, value)
  }

  selectByID (id: string): ?any {
    return this.selectByProp('testID', id)
  }
}

export default function explore (component: any): ComponentExplorer {
  return new ComponentExplorer(component)
}
