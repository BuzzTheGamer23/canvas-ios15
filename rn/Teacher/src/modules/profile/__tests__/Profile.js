/* @flow */
const { it, expect } = global
import 'react-native'
import React from 'react'
import Profile from '../Profile.js'

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer'

import explore from '../../../utils/explore'

it('renders correctly', () => {
  const tree = renderer.create(
    <Profile />
  ).toJSON()

  expect(tree).toMatchSnapshot()

  const view = explore(tree).selectByID('module.profile') || {}
  expect(view.props.accessible).toBeTruthy()
})
