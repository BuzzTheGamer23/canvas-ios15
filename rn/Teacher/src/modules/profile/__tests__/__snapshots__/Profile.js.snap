// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  accessible={true}
  testID="module.profile"
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    This is the profile
  </Text>
  <View
    accessibilityComponentType="button"
    accessibilityLabel={undefined}
    accessibilityTraits={
      Array [
        "button",
      ]
    }
    accessible={true}
    hitSlop={undefined}
    isTVSelectable={true}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#0C42FD",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Logout
      </Text>
    </View>
  </View>
</View>
`;
